### ISSUE TITLE
Config file paths might be better as relative

### ISSUE THREAD

**Issue by twVolc** [twVolc/PyCamPermanent#151]
Happy to be told otherwise on this, but I think it would be more suitable to save all of the paths that are to files saved alongside the config file (i.e., `pcs_lines`, `dil_lines`, `image_registration`, `default_cam_geom`) as relative paths; other ones, like `img_dir` should stay as absolute paths. This is just because I sometimes move the processing folder, or rename it, to make things clearer for me, but in doing this I break the config file within it. It makes more sense to me to have that whole folder independent of location - you could move it around but it would still work as long as all the files within the folder remain together.

**Comment by ubdbra001** [comment]
Had a bit of a think about this today. How does this sound:

__Saving:__
- Specify list of params that are paths
- Try setting the value of each param to be relative to the saving dir (where the config file will be saved)
- If this involves moving up to a parent directory, then save that as an absolute path.

__Loading:__
- On start-up:
  - Is the path to the default config file relative or absolute?
    - If relative: then use the paths in the config file unmodified
    - If absolute: then check the path being loaded
      - If relative: then prepend the dir where the config file is located
      - If absolute: then use as is
- On loading new config file:
  - Check the path being loaded from the config file:
    - If relative: then prepend the dir where the config file is located
    - If absolute: then use as is

I think those rules should cover all cases. Could you take a look @twVolc and tell me what you think?

**Comment by ubdbra001** [comment]
This would link to #141 as during loading there can be a check that the file exists, and if not it can raise an exception that gets passed to the GUI to show an error message

**Comment by twVolc** [comment]
Just thinking out loud about the initial installation - all paths at that point need to be relative. Will any of this change interfere with that? I guess not as the default config stays as it is? What we might see happen is that paths that are originally relative, when saved in a new config file, even if the path hasn't changed, will now be written as absolute if they aren't part of the config folder group. But I don't think this will cause any issues in any way? In other words, any paths that are relative to pycam can just become absolute paths as we're assuming the pycam installation stays in the same place (I think this is a fair enough assumption), so whenever we're talking about a relative path we're only ever talking about a path relative to the config file. As I write this I think I'm actually confusing myself, as the original install needs to know that all the relative paths are relative to the installation in that instance.... Just had another thought - what if when saving config files we just save the paths as a word like `None` if the files are in the config folder. When loading in the files if the path is simply that string we know to get the path from the config file path. The only slight edge case incompatibility is if randomly a directory path is literally named `None` and then this breaks, but this really shouldn't happen. Alternatively make the path empty - no value after the key in the config file - this means just grab the file from the config - you would then just need a list of what the respective file is called when saved (i.e. the image registration file is allways `img_reg` with iether `.pkl` or `'npy' depeding on the registration type; all other files always have exactly the same name I believe.

Checking if moving up to a parent directory sounds fine to me too, so I think whichever seems more optimal/logical probably works? I think in my suggestion the relative paths still stay relative to the pycam install, and the files relative to the config file will just be found out as you go along.

Another alternative might be having some kind of dictionary that has all the attributes as keys and has a value of something like `conf` if it's always saved in the config folder. That way when loading we can just check which attributes need to have their paths adjusted to be in the same folder as the config file.

**Comment by ubdbra001** [comment]
> Will any of this change interfere with that? I guess not as the default config stays as it is?

I don't think so, this is why it checks to see if the path to the default config is relative on start up. My thinking is that only the default supplied with pycam will be relative (needs to be as we don't know where a user is going to put the pycam folder). When you set your own default that will be absolute.

> What we might see happen is that paths that are originally relative, when saved in a new config file, even if the path hasn't changed, will now be written as absolute if they aren't part of the config folder group. But I don't think this will cause any issues in any way? In other words, any paths that are relative to pycam can just become absolute paths as we're assuming the pycam installation stays in the same place (I think this is a fair enough assumption)

Yes, but I was thinking a bit broader than that. Effectively: anything in the same directory as the config file gets a relative path (so it is relative to the config file), anything else gets an absolute path. You could make a relative path for the latter files (e.g. `../../Projects/PyCamPermanent/etc`), but if you move the dir with the config file somewhere else then it would likely break them, so not really worth it.

> what if when saving config files we just save the paths as a word like None if the files are in the config folder. When loading in the files if the path is simply that string we know to get the path from the config file path. The only slight edge case incompatibility is if randomly a directory path is literally named None and then this breaks, but this really shouldn't happen. Alternatively make the path empty - no value after the key in the config file - this means just grab the file from the config - you would then just need a list of what the respective file is called when saved (i.e. the image registration file is allways img_reg with iether .pkl or `'npy' depeding on the registration type; all other files always have exactly the same name I believe.

Not a fan of this idea. I don't think it would make addressing this issue any easier and obfuscates the specific file that the software is looking for, plus I don't think having a config file with a bunch of `None`s in it is very transparent for the end user. Happy to discuss it more if you think there's a good reason to do it this way that I'm missing.

> Another alternative might be having some kind of dictionary that has all the attributes as keys and has a value of something like conf if it's always saved in the config folder. That way when loading we can just check which attributes need to have their paths adjusted to be in the same folder as the config file.

Not sure I fully understand what you're saying here.

**Comment by twVolc** [comment]
Ok I think I'm probably overcomplicating things, so if you're confident that your solution will work I'd say just go with that. If we encounter issues we can debug as we go.

**Comment by ubdbra001** [comment]
Okay, will do.
I have discovered the first fly in the ointment of my original plan: some of the path parameters in the config file are just strings while other are lists of strings (e.g. `pcs_lines`).
Not insurmountable, but I'm sure there's a more elegant way of handling it than my first attempt!
